File: scripts/reference-solutions/palindrome-pairs.py
import sys


def main():
    lines = sys.stdin.read().split("\n")
    n = int(lines[0])
    words = [lines[1 + i].rstrip("\r") for i in range(n)]
    pos = {w: i for i, w in enumerate(words)}

    def is_pal(s):
        return s == s[::-1]

    res = set()
    for i, w in enumerate(words):
        length = len(w)
        for k in range(length + 1):
            prefix = w[:k]
            suffix = w[k:]
            # prefix palindrome -> words[j] + w is a palindrome, j = reverse(suffix)
            if is_pal(prefix):
                r = suffix[::-1]
                if r in pos and pos[r] != i:
                    res.add((pos[r], i))
            # suffix palindrome -> w + words[j] is a palindrome, j = reverse(prefix)
            if is_pal(suffix):
                r = prefix[::-1]
                if r in pos and pos[r] != i:
                    res.add((i, pos[r]))
    if not res:
        print("EMPTY")
    else:
        print("\n".join(f"{a} {b}" for a, b in sorted(res)))


main()

File: scripts/reference-solutions/palindrome-pairs.cpp
#include <algorithm>
#include <array>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <set>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

static bool isPalindrome(const string& value) {
  int left = 0;
  int right = static_cast<int>(value.size()) - 1;
  while (left < right) {
    if (value[left++] != value[right--]) {
      return false;
    }
  }
  return true;
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int n;
  cin >> n;
  string line;
  getline(cin, line);

  vector<string> words(n);
  unordered_map<string, int> pos;
  pos.reserve(n * 2 + 1);
  for (int i = 0; i < n; ++i) {
    getline(cin, words[i]);
    if (!words[i].empty() && words[i].back() == '\r') {
      words[i].pop_back();
    }
    pos[words[i]] = i;
  }

  set<pair<int, int>> result;
  for (int i = 0; i < n; ++i) {
    const string& word = words[i];
    for (size_t cut = 0; cut <= word.size(); ++cut) {
      string prefix = word.substr(0, cut);
      string suffix = word.substr(cut);

      if (isPalindrome(prefix)) {
        string reversed = suffix;
        reverse(reversed.begin(), reversed.end());
        auto it = pos.find(reversed);
        if (it != pos.end() && it->second != i) {
          result.insert({it->second, i});
        }
      }

      if (isPalindrome(suffix)) {
        string reversed = prefix;
        reverse(reversed.begin(), reversed.end());
        auto it = pos.find(reversed);
        if (it != pos.end() && it->second != i) {
          result.insert({i, it->second});
        }
      }
    }
  }

  if (result.empty()) {
    cout << "EMPTY\n";
    return 0;
  }
  for (auto [left, right] : result) {
    cout << left << ' ' << right << '\n';
  }
  return 0;
}
